### PR TITLE
feat(AutoTLS): opt-in WSS certs from p2p-forge at libp2p.direct

### PR DESCRIPTION
**Pull request by aschmahmann**
This PR is WIP but the intention is to allow any publicly reachable node on the network to be reachable via Secure WebSockets and thereby improve reachability from browsers. For anyone who reads this and is unfamiliar with this work, don't worry plenty of documentation to come before we're ready to merge.

## TLDR

```console
$ ipfs config --json AutoTLS.Enabled true
$ export GOLOG_LOG_LEVEL="error,autotls=debug,p2p-forge/client=debug" # to see whats happening
$ ipfs daemon
```

## TODO 

Some outstanding items here are:
- [ ] ~the libp2p.direct service needs to be ready for production use~ tracked in https://github.com/ipfs/kubo/issues/10560
- [x]  https://github.com/libp2p/go-libp2p/pull/2854 lands
- [x] documentation
- [x] need to be happy with the config file syntax
- [x] remove debugging logs associated with acquiring certificates
- [x] we should probably wait until we have confidence in our external addresses before trying to get a certificate
- [ ] ~multiplex WSS on the same port as TCP (e.g. by muxing on the ALPN in the TLS cert) so people don't need to open new ports~
  - descoped from this PR, will land once https://github.com/libp2p/go-libp2p/pull/2984 is ready
  - tracked in https://github.com/ipfs/kubo/issues/10560
- [x] there's currently an issue where the application doesn't realize it's acquired a certificate and that it's ready for use (happens at least on first acquisition)
- [ ] figure out what's going on with excessive logs like
   - [ ]  "http: TLS handshake error from <ip:port>: EOF"
     - unable to reproduce anymore, assuming fixed
   - [x] `ERROR        p2p-forge/client.maintenance        certmagic@v0.21.4/maintain.go:521        error while checking storage for updated ARI; updating ARI now` (from `ipfs` at `collab-cluster-sv15-1`)
- [x] remove `replace github.com/libp2p/go-libp2p`  from all three `go.mod`

### CODE REVIEW

File: config/forge.go
Comment (lidel) [inline config/forge.go:1]: @aschmahmann i've moved configuration to struct here, everything optional
tldr enabling via `ipfs config --json Swarm.ForgeClient.Enabled true`

File: config/forge.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	Enabled Flag `json:",omitempty"`
+
+	// Optional override of the parent domain that will be used
+	ForgeDomain *OptionalString `json:",omitempty"`
+
+	// Optional override of HTTP API that acts as ACME DNS-01 Challenge broker
+	ForgeEndpoint *OptionalString `json:",omitempty"`
+
+	// Optional Authorization token, used with private/test instances of p2p-forge
+	ForgeAuth *OptionalString `json:",omitempty"`
```
Comment (lidel) [inline config/forge.go:19]: TODO: this is not wired-up yet (`FORGE_ACCESS_TOKEN` should override this config)

File: core/node/libp2p/addrs.go [hunk 2]
```diff
@@ -107,3 +130,41 @@ func ListenOn(addresses []string) interface{} {
...
+	const authForgeHeader = "Forge-Authorization"
+	authKey, foundAuthKey := os.LookupEnv(authEnvVar)
+
+	certMgr, err := p2pforge.NewP2PForgeCertMgr(
+		p2pforge.WithModifiedForgeRequest(func(req *http.Request) error {
+			if foundAuthKey {
+				req.Header.Set(authForgeHeader, authKey)
+			}
+			return nil
+		}),
```
Comment (lidel) [inline core/node/libp2p/addrs.go:150]: TODO: move to `p2p-forge/client` library, and call more specific funcs here:
- `WithForgeAuth(authKey)`
- `WithUserAgent(GetUserAgentVersion())`

File: config/forge.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+// browser clients. More: https://github.com/ipshipyard/p2p-forge#readme
+type ForgeClient struct {
+	// Enables the p2p-forge feature
+	Enabled Flag `json:",omitempty"`
+
+	// Optional override of the parent domain that will be used
+	ForgeDomain *OptionalString `json:",omitempty"`
+
+	// Optional override of HTTP API that acts as ACME DNS-01 Challenge broker
+	ForgeEndpoint *OptionalString `json:",omitempty"`
```
Comment (lidel) [inline config/forge.go:16]: TODO: wire up and use WithDefault from consts here

File: core/node/libp2p/addrs.go [hunk 4]
```diff
@@ -107,3 +131,40 @@ func ListenOn(addresses []string) interface{} {
...
 }
+
+func P2PForgeCertMgr(cfg config.ForgeClient) interface{} {
+	return func() (*p2pforge.P2PForgeCertMgr, error) {
+		storagePath, err := config.Path("", "p2p-forge-certs")
+		if err != nil {
+			return nil, err
+		}
+
+		certmagic.Default.Logger = logging.Logger("certmagic").Desugar()
```
Comment (lidel) [inline core/node/libp2p/addrs.go:142]: @aschmahmann thoughts on making this `p2p-forge/certmagic` ?

```suggestion
		certmagic.Default.Logger = logging.Logger("p2p-forge/certmagic").Desugar()
```

**Comment by lidel** [comment]
Rebased to use the latest commit from go-libp2p master branch that includes https://github.com/libp2p/go-libp2p/pull/2854.

Go Lint fails due to unrelated deprecation that will be fixed once we have go-libp2p release with this.

**Comment by lidel** [comment]
Pushed:
- Docs
- Logger cleanup, should be up to release standard.
  - Debug logs can be enabled via `export GOLOG_LOG_LEVEL="error,p2p-forge/client=debug"` and by default only ERROR level is printed to stdout.

Next steps:
- My plan is to wait for go-libp2p [0.37 release](https://github.com/libp2p/go-libp2p/issues/2901), bump dependency and resolve issues
- merge this PR, and start testing  docker image from master branch in our collab cluster. this way we don't need to wait for kubo [0.32-rc1](https://github.com/ipfs/kubo/issues/10547), and get real world feedback faster.

**Comment by lidel** [comment]
Switched to go-libp2p 0.37 from master branch and resolved conflicts, I'll check on CI tomorrow.

File: docs/config.md [hunk 5]
```diff
@@ -1716,6 +1722,83 @@ Default: `true`
 
 Type: `flag`
 
+### `Swarm.ForgeClient`
+
+Configuration options for optional [p2p-forge] client that provides Kubo with DNS+TLS+WebSockets multiaddrs that allow for direct connections from JS running in web browsers.
+
+The default settings use instance at `libp2p.direct` and TLS certificates from Let's Encrypt.
```
Comment (2color) [inline docs/config.md:1729]:
```suggestion
By default, AutoTLS requests TLS certificates from Let's Encrypt and uses the `libp2p.direct` domain.
```

**Review by 2color — COMMENTED** [review]
As discussed, left some feedback on naming. I can submit more suggestions thorough PR tomorrow.

File: core/node/libp2p/addrs.go [hunk 6]
```diff
@@ -107,3 +131,47 @@ func ListenOn(addresses []string) interface{} {
...
+		certmagic.DefaultACME.Logger = forgeLogger
+
+		certMgr, err := p2pforge.NewP2PForgeCertMgr(
+			p2pforge.WithLogger(forgeLogger.Sugar()),
+			p2pforge.WithForgeDomain(cfg.DomainSuffix.WithDefault(config.DefaultDomainSuffix)),
+			p2pforge.WithForgeRegistrationEndpoint(cfg.RegistrationEndpoint.WithDefault(config.DefaultRegistrationEndpoint)),
+			p2pforge.WithCAEndpoint(cfg.CAEndpoint.WithDefault(config.DefaultCAEndpoint)),
+			p2pforge.WithForgeAuth(cfg.RegistrationToken.WithDefault(os.Getenv(p2pforge.ForgeAuthEnv))),
+			p2pforge.WithUserAgent(version.GetUserAgentVersion()),
+			p2pforge.WithCertificateStorage(&certmagic.FileStorage{Path: storagePath}))
```
Comment (lidel) [inline core/node/libp2p/addrs.go:157]: Found a bug in cluster:

- `Oct 26 07:24:33 collab-cluster-sv15-1 ipfs[99759]: 2024-10-26T07:24:33.169Z        ERROR        p2p-forge/client.maintenance        certmagic@v0.21.4/maintain.go:521        error while checking storage for updated ARI; updating ARI now        {"identifiers": ["*.k51qzi5uqu5dljgcjt37azmfvj1u9b1v9okfd0vhd4gkbcs75cmy0shqjoml6q.libp2p.direct"], "cert_hash": "ac1b9af9ec6e49e5e445593e02620a25a4f6bf5f8013ac0ed7a0206cb9c5f9bd", "ari_unique_id": "kydGmAOpUWiOmNbEQkjbI79YlNI.A_zqjUpBrD6AJHDUgsoK9qj6", "cert_expiry": "2025-01-23T11:57:29.000Z", "error": "loading cert metadata: open /home/ipfs/.local/share/certmagic/certificates/acme-v02.api.letsencrypt.org-directory/wildcard_.k51qzi5uqu5dljgcjt37azmfvj1u9b1v9okfd0vhd4gkbcs75cmy0shqjoml6q.libp2p.direct/wildcard_.k51qzi5uqu5dljgcjt37azmfvj1u9b1v9okfd0vhd4gkbcs75cmy0shqjoml6q.libp2p.direct.json: no such file or directory"}`
* `Oct 26 07:24:33 collab-cluster-sv15-1 ipfs[99759]: 2024-10-26T07:24:33.257Z        ERROR        p2p-forge/client.maintenance        certmagic@v0.21.4/maintain.go:181        updating ARI        {"error": "got new ARI from acme-v02.api.letsencrypt.org-directory, but failed loading stored certificate metadata: loading cert metadata: open /home/ipfs/.local/share/certmagic/certificates/acme-v02.api.letsencrypt.org-directory/wildcard_.k51qzi5uqu5dljgcjt37azmfvj1u9b1v9okfd0vhd4gkbcs75cmy0shqjoml6q.libp2p.direct/wildcard_.k51qzi5uqu5dljgcjt37azmfvj1u9b1v9okfd0vhd4gkbcs75cmy0shqjoml6q.libp2p.direct.json: no such file or directory"}`

I suspect this is another race condition where certmanager things (like location `.local/share/certmagic/`) are initialized with default config before our config is applied (same problem as the logger). 

Will look into this.

Comment (lidel) [reply core/node/libp2p/addrs.go]: Potential fix in https://github.com/ipfs/kubo/pull/10521/commits/e51d907854be63abb4a112abf72540b436f7ffb2 and https://github.com/ipshipyard/p2p-forge/commit/f599f4803109aeaeae1c80756580759889a92e16. It should remove surface for race condition in setting storage paths AFTER maintenance job is started.

 I'm going to deploy to collab cluster and see if ARI error is gone.

File: docs/changelogs/v0.32.md [hunk 7]
```diff
@@ -15,6 +15,12 @@
 
 ### 🔦 Highlights
 
+#### 🎯 Opt-in `/wss` Certificates via libp2p.direct
+
+This release introduces an experimental feature that significantly improves how browsers can connect to Kubo node.
+Opt-in configuration allows Kubo node to obtain trusted certificates for Secure WebSocket (WSS) connections without manual intervention.
```
Comment (2color) [inline docs/changelogs/v0.32.md:21]:
```suggestion
Opt-in configuration allows Kubo nodes to obtain CA signed TLS certificates for Secure WebSocket (WSS) connections automatically.
```

File: docs/changelogs/v0.32.md [hunk 8]
```diff
@@ -15,6 +15,12 @@
 
 ### 🔦 Highlights
 
+#### 🎯 Opt-in `/wss` Certificates via libp2p.direct
```
Comment (2color) [inline docs/changelogs/v0.32.md:18]:
```suggestion
#### 🎯 AutoTLS: Automatic Certificates for `/wss` via libp2p.direct
```

File: docs/config.md [hunk 9]
```diff
@@ -1716,6 +1722,86 @@ Default: `true`
...
+internet) to automatically obtain a wildcard TLS certificate for a DNS name
+unique to their PeerID at `*.[PeerID].libp2p.direct`. This enables direct
+libp2p connections and retrieval of IPFS content from browsers using Secure
+WebSockets, without requiring manual domain registration and configuration.
+
+Under the hood, `libp2p.direct` acts as an [ACME DNS-01 Challenge](https://letsencrypt.org/docs/challenge-types/#dns-01-challenge)
+broker for obtaining these wildcard TLS certificates.
+
+By default, the certificates are requested from Let's Encrypt.
+Origin and rationale for this project can be found in [community.letsencrypt.org discussion].
```
Comment (2color) [inline docs/config.md:1737]: Missing URL

File: docs/config.md [hunk 10]
```diff
@@ -1716,6 +1722,86 @@ Default: `true`
...
+[community.letsencrypt.org discussion]: https://community.letsencrypt.org/t/feedback-on-raising-certificates-per-registered-domain-to-enable-peer-to-peer-networking/223003
+
+Default: `{}`
+
+Type: `object`
+
+#### `Swarm.AutoTLS.Enabled`
+
+> [!CAUTION]
+> This is an EXPERIMENTAL feature and should not be used in production yet.
```
Comment (2color) [inline docs/config.md:1754]: Should we create a discussion/issue specifically for feedback and link it here?

File: docs/config.md [hunk 11]
```diff
@@ -1716,6 +1722,86 @@ Default: `true`
...
+> Feel free to enable it and [report issues](https://github.com/ipfs/kubo/issues/new/choose) if you want to help with testing.
+
+Enables **EXPERIMENTAL** [p2p-forge] client. This feature works only if your Kubo node is publicly diallable.
+
+If enabled, it will detect when `.../tls/sni/.../ws` is present in [`Addresses.Swarm`](#addressesswarm)
+and SNI is matching `Swarm.AutoTLS.DomainSuffix`, and set up a trusted TLS certificate matching the domain name used in Secure WebSockets (WSS) listener.
+
+If you want to test this, add `/ip4/0.0.0.0/tcp/4082/tls/sni/*.libp2p.direct/ws` to [`Addresses.Swarm`](#addressesswarm).
```
[the remark this answers is not in the document]
Comment (lidel) [reply docs/config.md:1762]: Not yet, we need https://github.com/libp2p/go-libp2p/pull/2984 for that. 

The idea is that TCP port sharing will land in go-libp2p before we enable AutoTLS by default.
For opt-in, using separate port is fine, just need to remember to remind people that they need to safelist additional port on their firewall.

File: core/node/libp2p/addrs.go (hunk 6 above)
Comment (lidel) [reply core/node/libp2p/addrs.go]: Ok, the error seems to be gone. Considering this resolved, but keeping this open for visibility.

File: docs/config.md (hunk 10 above)
Comment (lidel) [reply docs/config.md]: Created https://github.com/ipfs/kubo/issues/10560. 

I also moved `AutoTLS` to the top level of config file, just to future proof it. Rationale in https://github.com/ipfs/kubo/pull/10521/commits/73c3fd3d15d0ca9cd74aaea770166fa12e40bb2e

**Review by lidel — APPROVED** [review]
We've been running this in collab cluster and no new errors were found. 
Configuration and docs are in place. This is likely as good as it can be without asking wider community for feedback. 

Merging to include it in tomorrow's Kubo 0.32.0-rc1 (https://github.com/ipfs/kubo/issues/10547)
